uniffi/src/ffi/rustcalls.rs

```diff
@@ -0,0 +1,251 @@
+    // error_buf is MaybeUninit to avoid dropping the value that the consumer code sends in:
+    //   - Consumers should send in a zeroed out RustBuffer.  In this case dropping is a no-op and
+    //     avoiding the drop is a small optimization.
+    //   - If consumers pass in invalid data, then we should avoid trying to drop in.  In
+    //     particular, we don't want to try to free any data the consumer has allocated.
+}
+
+#[allow(dead_code)]
+const CALL_SUCCESS: i8 = 0; // CALL_SUCCESS is set by the calling code
+const CALL_ERROR: i8 = 1;
+const CALL_PANIC: i8 = 2;
+
+// A trait for errors that can be thrown to the FFI code
+//
+// This gets implemented in uniffi_bindgen/src/scaffolding/templates/ErrorTemplate.rs
+pub trait FfiError: RustBufferViaFfi {}
+
+// Generalized rust call handling function
+// callback should map errors to a RustBuffer that contains them
+fn make_call<F, R>(out_status: &mut RustCallStatus, callback: F) -> R::Value
+where
+    F: panic::UnwindSafe + FnOnce() -> Result<R, RustBuffer>,
+    R: IntoFfi,
+{
+    let result = panic::catch_unwind(|| {
+        // We should call:
+        //
+        // init_panic_handling_once();
+        //
+        // This is called in ffi-support::call_with_result_impl().  The current plan is to make it
+        // `pub` in ffi-support and call it from here.
+        callback().map(|v| v.into_ffi_value())
+    });
+    match result {
+        // Happy path.  Note: no need to update out_status in this case because the calling code
+        // initializes it to CALL_SUCCESS
+        Ok(Ok(v)) => v,
+        // Callback returned an Err.
+        Ok(Err(buf)) => {
+            out_status.code = CALL_ERROR;
+            unsafe {
+                out_status.error_buf.as_mut_ptr().write(buf);
+            }
+            R::ffi_default()
+        }
+        // Callback panicked
+        Err(cause) => {
+            out_status.code = CALL_PANIC;
+            // Try to coerce the cause into a RustBuffer containing a String.  Since this code can
+            // panic, we need to use a second catch_unwind().
+            let message_result = panic::catch_unwind(panic::AssertUnwindSafe(move || {
+                // The documentation suggests that it will *usually* be a str or String.
+                let message = if let Some(s) = cause.downcast_ref::<&'static str>() {
+                    (*s).to_string()
+                } else if let Some(s) = cause.downcast_ref::<String>() {
+                    s.clone()
+                } else {
+                    "Unknown panic!".to_string()
+                };
+                log::error!("Caught a panic calling rust code: {:?}", message);
+                String::lower(message)
+            }));
+            if let Ok(buf) = message_result {
+                unsafe {
+                    out_status.error_buf.as_mut_ptr().write(buf);
+                }
+            }
+            // Ignore the error case.  We've done all that we can at this point
```

Comment:
In practice, I guess this will cause the foreign-language bindings to just read an empty string from the (presumably) empty buffer that it passed in. May be worth noting that explicitly here.